pkg/web_app/lib/src/widget/widget.dart

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:js_interop';
+
+import 'package:collection/collection.dart';
+import 'package:web/web.dart';
+
+import '../web_util.dart';
+import 'completion/widget.dart' deferred as completion;
+
+/// Widget entry
+typedef _WidgetEntry = ({
+  /// Name of the widget referenced in `data-widget="<name>"`.
+  String name,
+
+  /// Function to create an instance of the widget given an element and options.
+  ///
+  /// [element] which carries `data-widget="$name"`.
+  /// [options] a map from options to values, where options are specified as
+  /// `data-$name-$option="$value"`.
+  ///
+  /// Hence, a widget called `completion` is created on an element by adding
+  /// `data-widget="completion"`. And option `src` is specified with:
+  /// `data-completion-src="$value"`.
+  FutureOr<void> Function(Element element, Map<String, String> options) create,
+
+  /// Load widget library, if deferred (otherwise this can be `null`).
+  Future<void> Function()? loadLibrary,
+});
+
+// List of registered widgets
+final _widgets = <_WidgetEntry>[
+  (
+    name: 'completion',
+    create: completion.create,
+    loadLibrary: completion.loadLibrary,
+  ),
+];
+
+Future<void> setupWidgets() async {
```

Comment:
I'm making it `void` because I don't want people to await it :D

`setupWidgets` does what it does, you don't need to know if it failed, it'll report errors in console.